File: examples/pingpong/scala.rb
#!/usr/bin/env ruby

# cf. http://lamp.epfl.ch/~phaller/doc/ActorsTutorial.html

$:.push File.join( File.dirname(__FILE__), "..", "..", "lib" )

require 'dramatis/actor'

class Ping
  include Dramatis::Actor
  def initialize times, pong
    @pings_left = times
    cast( pong ).ping actor.name 
  end
  def pong caller
    if @pings_left % 1000 == 0
      puts "Ping: pong"
    end
    if @pings_left > 0
      @pings_left -= 1
      cast( caller ).ping actor.name
    end
  end
end

class Pong
  include Dramatis::Actor
  def initialize
    @pong_count = 0
  end
  def ping caller
    if @pong_count % 1000 == 0
      puts "Pong: ping #{@pong_count}"
    end
    @pong_count += 1
    cast( caller ).pong actor.name
  end
end

pong = Pong.new
ping = Ping.new 10000, pong

File: examples/pingpong/scala.py
#!/usr/bin/env python

# cf. Scala by Example, Chapter 3

import time
import random
import inspect
import sys
import os.path

sys.path[0:0] = [ os.path.join( os.path.dirname( inspect.getabsfile( inspect.currentframe() ) ), '..', '..', 'lib' ) ]

from logging import warning

import dramatis

class Ping (dramatis.Actor):

    def __init__(self,times,pong):
        self.pings_left = times
        dramatis.release( pong ).ping( self )

    def pong(self,caller):
        if self.pings_left % 1000 == 0:
            print "Ping: pong"
        if self.pings_left > 0:
            self.pings_left -= 1
            dramatis.release( caller ).ping( self )
            
class Pong (dramatis.Actor):

    def __init__(self):
        self.pong_count = 0

    def ping(self,caller):
        if self.pong_count % 1000 == 0:
            print "Pong: ping", self.pong_count
        self.pong_count += 1
        dramatis.release( caller ).pong( self )

pong = Pong()
ping = Ping( 1000, pong )
